File: Data Structures & Algorithms/three-integer-sum/submission-2.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        int[] sortedNums = nums.clone();
    Arrays.sort(sortedNums);
    int l;int r;
    List<List<Integer>> result = new ArrayList<>();
    for(int i=0;i<nums.length;++i) {
        if (i > 0 && sortedNums[i] == sortedNums[i - 1]) {
            continue;
        }
        l = i + 1;
        r = nums.length - 1;
        int target = -sortedNums[i];
        while (l < r) {
            if (sortedNums[l] + sortedNums[r] < target) ++l;
            else if (sortedNums[l] + sortedNums[r] > target) --r;
            else {
                result.add(Arrays.asList(-target, sortedNums[l], sortedNums[r]));
                ++l;
                --r;
                while (l < r && sortedNums[l] == sortedNums[l - 1]) l++;
                while (l < r && sortedNums[r] == sortedNums[r + 1]) r--;
            }
        }
    }
    return result;
    }
}

File: Data Structures & Algorithms/search-2d-matrix/submission-2.cpp
class Solution {
public:
    static bool searchMatrix(std::vector<std::vector<int>>& matrix, int target) {
        int l=0;
        int r=matrix.size()*matrix[0].size()-1;
        int mid = l + (r-l)/2;

        while (l<=r)
        {
            const int value = matrix[mid/matrix[0].size()][mid%matrix[0].size()];
            if (target < value) r = mid-1;
            else if (target > value) l=mid+1;
            else return true;
            mid = l + (r-l)/2;
        }
        return false;
    }
};